include/v8-instance.h

```diff
@@ -29,53 +36,24 @@ class v8_instance {
         });
     }
 
-    bool run_instance() {
-        is_finish.store(false);
-        v8::Locker locker(isolate);
-        v8::Isolate::Scope isolate_scope(isolate);
-        v8::HandleScope handle_scope(isolate);
-        v8::TryCatch try_catch(isolate);
-        v8::Local<v8::Context> local_ctx = v8::Local<v8::Context>::New(isolate, context);
-        v8::Context::Scope context_scope(local_ctx);
-
-        const int argc = 1;
-        auto array = v8::ArrayBuffer::New(isolate, store);
-        v8::Local<v8::Value> argv[argc] = { array };
-        v8::Local<v8::Value> result;
-
-        v8::Local<v8::Function> local_function = v8::Local<v8::Function>::New(isolate, function);
-        if (!local_function->Call(local_ctx, local_ctx->Global(), argc, argv).ToLocal(&result)) {
-            v8::String::Utf8Value error(isolate, try_catch.Exception());
-            std::cout << "Can not run script: " << std::string(*error, error.length()) << std::endl;
-            is_finish.store(true);
-            return false;
-        }
-
-        is_finish.store(true);
-        return true;
-    }
-
-    void wrap_external_memory(char* data_ptr, size_t size) {
-        store = v8::ArrayBuffer::NewBackingStore(data_ptr, size, v8::BackingStore::EmptyDeleter, nullptr);
+    seastar::future<bool> run_instance(v::ThreadPool& thread_pool, int timeout, std::span<char> data) {
+        return seastar::with_semaphore(mtx, 1, [this, &thread_pool, timeout, data](){
+            is_cancel = false;
+            watchdog.rearm(seastar::lowres_clock::time_point(seastar::lowres_clock::now() + std::chrono::seconds(timeout)));
+            return thread_pool.submit([this, data](){
+                run_instance_internal(data);
+            })
+            .then([this] {
+                if (!is_cancel) {
+                    watchdog.cancel();
+                }
+        	    return seastar::make_ready_future<bool>(is_cancel);
+            });
+        });
     }
 
-    void stop_execution_loop(std::chrono::time_point<std::chrono::high_resolution_clock> start_time, double timeout) {
-        while (true) {
-            if (is_finish.load()) {
-                break;
-            }
-
-            auto now = std::chrono::high_resolution_clock::now();
-            std::chrono::duration<double> time = now - start_time;
-            if (time.count() > timeout) {
-                //This method can be used by any thread even if that thread has not
-                //acquired the V8 lock with a Locker object.
-                isolate->TerminateExecution();
-                break;
-            } else {
-                std::this_thread::yield();
-            }
-        }
+    void stop_execution_loop() {
+        isolate->TerminateExecution();
```

Comment:
maybe add some check that protects from second call? like if the instance is already stopped then crash the whole thing altogether

Reply:
Done